src/util/dynamicfilesequencedownloader.cpp

```diff
@@ -0,0 +1,636 @@
+        }
+        catch (const ghoul::RuntimeError& ex) {
+            LERROR(std::format("An error occurred: '{}'", ex.what()));
+        }
+
+        if (!success) {
+            if (attempt < MaxRetries) {
+                LINFO(std::format("Retry nr {}.", attempt + 1));
+                std::this_thread::sleep_for(std::chrono::seconds(1));
+
+                response.start();
+                response.wait();
+            }
+            else {
+                LERROR(std::format(
+                    "Failed according to warning above with HTTP request of URL: {}",
+                    httpDataRequest
+                ));
+            }
+        }
+        attempt++;
+    }
+
+    if (!success) {
+        return;
+    }
+
+    int index = 0;
+    for (const nlohmann::json& element : jsonResult["files"]) {
+        std::string timestamp = element["timestamp"].get<std::string>();
+        std::string url = element["url"].get<std::string>();
+
+        //timestamp = "2022-11-13T16:14:00.000";
+        //url =
+        //https://iswaA-webservice1.ccmc.gsfc.nasa.gov/
+        //...iswa_data_tree/model/solar/WSA5.X/fieldlines/
+        //...GONG_Z/trace_pfss_intoout/2022/11/2022-11-13T16-14-00.000.osfls";
+
+        std::string fileName = url.substr(url.find_last_of("//"));
+        std::filesystem::path destination = _syncDir;
+        destination += fileName;
+
+        double time = Time::convertTime(timestamp);
+        File fileElement = {
+            .timestep = timestamp,
+            .time = time,
+            .URL = url,
+            .path = destination,
+            .cadence = 0,
+            .availableIndex = index
+        };
+        if (std::filesystem::exists(destination)) {
+            fileElement.download = nullptr;
+            fileElement.state = File::State::Downloaded;
+            _downloadedFiles.push_back(destination);
+            trackFinishedDownloads(_trackSynced, destination);
+        }
+        else {
+            fileElement.download = std::make_unique<HttpFileDownload>(url, destination);
+            fileElement.state = File::State::Available;
+        }
+        _availableData.push_back(std::move(fileElement));
+        ++index;
+    }
+
+    const double cadence = calculateCadence();
+    for (File& element : _availableData) {
+        element.cadence = cadence;
+    }
+}
+
+double DynamicFileSequenceDownloader::calculateCadence() const {
+    double averageTime = 0.0;
+    if (_availableData.size() < 2) {
+        // If 0 or 1 files there is no cadence
+        return averageTime;
+    }
+    double time1 = Time::convertTime(_availableData.begin()->timestep);
+    double timeN = Time::convertTime(_availableData.rbegin()->timestep);
+    averageTime = (timeN - time1) / _availableData.size();
+
+    return averageTime;
+}
+
+void DynamicFileSequenceDownloader::downloadFile() {
+    ZoneScoped;
+
+    if (_filesCurrentlyDownloading.size() < 4 && !_queuedFilesToDownload.empty()) {
```

Comment:
Why `4`? In this case it would be better to place that into a `constexpr int xx = 4;` variable with a descriptive name